save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/controllers/DownloadProjectController.kt

```diff
@@ -168,69 +178,94 @@ class DownloadProjectController(private val configProperties: ConfigProperties)
+        } ?: require(executionType == ExecutionType.GIT) { "Test suites are not provided, but should for executionType=$executionType" }
+
         val execution = Execution(project, LocalDateTime.now(), LocalDateTime.now(),
-            ExecutionStatus.PENDING, "1", resourcesRootPath, 0, configProperties.executionLimit, executionType)
-        var execId: Long
-        log.debug("Knock-Knock Backend")
-        makeRequest(BodyInserters.fromValue(execution), "/createExecution") { it.bodyToMono(Long::class.java) }
-            .doOnNext { executionId ->
-                execId = executionId
-                makeRequest(BodyInserters.fromValue(testSuitesDto ?: getAllTestSuites(project, propertiesRelativePath)), "/saveTestSuites") { it.bodyToMono<List<TestSuite>>() }
-                    .doOnNext { testSuiteList ->
-                        makeRequest(BodyInserters.fromValue(getAllTests(resourcesRootPath, testSuiteList)), "/initializeTests?executionId=$executionId") { it.toBodilessEntity() }
-                            .doOnNext {
-                                // Post request to orchestrator to initiate its work
-                                log.debug("Knock-Knock Orchestrator")
-                                webClientOrchestrator
-                                    .post()
-                                    .uri("/initializeAgents")
-                                    .body(BodyInserters.fromValue(execution.also { it.id = execId }))
-                                    .retrieve()
-                                    .toEntity(HttpStatus::class.java)
-                                    .subscribe()
-                            }.subscribe()
-                    }.subscribe()
-            }.subscribe()
+            ExecutionStatus.PENDING, "ALL", resourcesRootRelativePath, 0, configProperties.executionLimit, executionType)
+        webClientBackend.makeRequest(BodyInserters.fromValue(execution), "/createExecution") { it.bodyToMono<Long>() }
+            .flatMap { executionId ->
+                val testResourcesRootAbsolutePath = getTestResourcesRootAbsolutePath(propertiesRelativePath, resourcesRootRelativePath)
+                discoverAndSaveTestSuites(project, testResourcesRootAbsolutePath, propertiesRelativePath)
+                    .flatMap { testSuites ->
+                        initializeTests(testSuites, testResourcesRootAbsolutePath, executionId)
+                            .then(initializeAgents(execution, executionId))
+                    }
+            }
+            .subscribe()
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun getTestResourcesRootAbsolutePath(propertiesRelativePath: String,
+                                                 resourcesRootRelativePath: String): String {
+        val saveProperties: SaveProperties = decodeFromPropertiesFile(
```

Comment:
What if there isn't `save.properties` file?

Reply:
Then, the `FileNotFoundException` will be thrown and the job in subscription will be stopped. I believe, I should also mark execution as failed on such errors

Reply:
Added a `doOnError` block in mono